### PR TITLE
[iOS] Programmatic focus triggered while evaluating user script should present the software keyboard

### PR DESCRIPTION
**Pull request by whsieh**
#### 2b0c968051a7552d2f54d39bf790e0970fce259d
<pre>
[iOS] Programmatic focus triggered while evaluating user script should present the software keyboard
<a href="https://bugs.webkit.org/show_bug.cgi?id=243416">https://bugs.webkit.org/show_bug.cgi?id=243416</a>
rdar://97945400

Reviewed by Aditya Keerthi and Devin Rousso.

Allow `-evaluateJavaScript:completionHandler:` and `-callAsyncJavaScript:…completionHandler:` to
set the `m_userIsInteracting` flag on WebPage during the scope of user script evaluation. This
allows apps (linked on or after certain versions of iOS) to use `element.focus()` to properly focus
text fields and other form controls, and show the software keyboard in the process.

Test: KeyboardInputTests.InputSessionWhenEvaluatingJavaScript

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:

Add a new SDK-aligned behavior flag to guard this behavior behind a linked-on-or-after check.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):

When executing user script (i.e. from the WebKit client), set the `m_userIsInteracting` flag to true
as long as the app is linked on or after the SDK versions corresponding to
`SDKAlignedBehavior::ProgrammaticFocusDuringUserScriptShowsInputViews`.

* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:

Add an API test to verify that using `-_evaluateJavaScriptWithoutUserGesture:…:` to evaluate script
that focuses a text field does not cause an input session to start; however, using either
`-callAsyncJavaScript:…:` or `-evaluateJavaScript:…:` starts an input session.

Canonical link: <a href="https://commits.webkit.org/253034@main">https://commits.webkit.org/253034@main</a>
</pre>

### CODE REVIEW

File: Source/WebKit/WebProcess/WebPage/WebPage.cpp [hunk 1]
```diff
@@ -3861,6 +3861,15 @@ void WebPage::runJavaScript(WebFrame* frame, RunJavaScriptParameters&& parameter
...
 
+    bool shouldAllowUserInteraction = [&] {
+#if PLATFORM(COCOA)
+        if (linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ProgrammaticFocusDuringUserScriptShowsInputViews))
+            return true;
+#endif
+        return m_userIsInteracting;
+    }();
+
+    SetForScope userIsInteractingChange { m_userIsInteracting, shouldAllowUserInteraction };
```
Comment (dcrousso) [inline Source/WebKit/WebProcess/WebPage/WebPage.cpp:3879]: do we also want to check `parameters.forceUserGesture`?  i think that `-[WKWebView _evaluateJavaScriptWithoutUserGesture:completionHandler:]` would probably not want to set the user interaction flag 😅

(part of me wants to ask you to move this to `ScriptController::executeScriptInWorld` since that's where the `UserGestureIndicator` is, but i understand that would involve lotsa `PageClient` plumbing so it's probably fine here)

Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/WebPage.cpp:3872]: This looks fine, since it doesn't appear that this method can get called while we're inside an existing `m_userIsInteracting` scope, but I wonder if we should be more defensive by making this:

```
SetForScope userIsInteractingChange { m_userIsInteracting, m_userIsInteracting | shouldAllowUserInteraction };
```

**Review by pxlcoder — APPROVED** [review]

File: Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.cpp [hunk 2]
```diff
@@ -172,6 +172,14 @@ static SDKAlignedBehaviors computeSDKAlignedBehaviors()
     if (linkedBefore(dyld_spring_2022_os_versions, DYLD_IOS_VERSION_15_4, DYLD_MACOSX_VERSION_12_3))
         disableBehavior(SDKAlignedBehavior::AuthorizationHeaderOnSameOriginRedirects);
 
+    if (linkedBefore(dyld_fall_2022_os_versions, DYLD_IOS_VERSION_16_0, DYLD_MACOSX_VERSION_13_0)) {
```
Comment (dcrousso) [inline Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.cpp]: i feel like this might be better as a separate change 😅

**Review by dcrousso — APPROVED** [review]

File: Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.cpp (hunk 2 above)
Comment (whsieh) [reply Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.cpp]: I can split it out, sure.

File: Source/WebKit/WebProcess/WebPage/WebPage.cpp (hunk 1 above)
Comment (whsieh) [reply Source/WebKit/WebProcess/WebPage/WebPage.cpp:3879]: > do we also want to check `parameters.forceUserGesture`?

Good catch! Yes, I'll check that flag as well (and add an API test to exercise this scenario).

> move this to `ScriptController::executeScriptInWorld`

Well, `m_userIsInteracting` is already a client-layer concept, so I don't think it would be an improvement to add all the client plumbing needed to make that work.

(I think it's also cleaner to be able to use `SetForScope`, as a variable on the stack, so we don't need to worry about setting and unsetting the bit)

**Comment by whsieh** [comment]
Thanks for the reviews!

**Comment by ivanrein** [comment]
Hi @whsieh @pxlcoder @dcrousso 
Do you know in which version of iOS is this released? I tried on ios simulator running ios 16 (xcode 14) and element.focus() still doesn't show keyboard.

**Comment by whsieh** [comment]
> Hi @whsieh @pxlcoder @dcrousso
> Do you know in which version of iOS is this released? I tried on ios simulator running ios 16 (xcode 14) and element.focus() still doesn't show keyboard.

Ah, so this change is not in iOS 16; we'll let you know when it lands in a developer beta.

Note also that this change is additionally behind a runtime flag, so once you update to a new SDK that contains this change, you'll need to rebuild your app against this new SDK version in order for `focus()` to pop open the software keyboard.

Thanks!

**Comment by sumeetp79** [comment]
Hi @whsieh @pxlcoder @dcrousso
Can you please let me know when the upcoming iOS release will include this change? It doesn't show up when element.focus() is called on the latest version of iPadOS, i.e. 15.7.1.
Thank you!

**Comment by AndreiBaicu26** [comment]
Hello! What is the latest release of WebKit where we can see this behavior implemented ? I see it working on iOS 14.1, but not working on iOS 16.1.  -- whent trying to focus() inside a setTimeout

Thanks :D

**Comment by whsieh** [comment]
> Hello! What is the latest release of WebKit where we can see this behavior implemented ? I see it working on iOS 14.1, but not working on iOS 16.1. -- whent trying to focus() inside a setTimeout
> 
> Thanks :D

I'm not sure why this behavior would work in iOS 14, but not iOS 16; in any case, the changes implemented in this patch aren't in the latest version of iOS 16 yet.

**Comment by AndreiBaicu26** [comment]
Thanks for the answer, found this jsFiddle http://jsfiddle.net/hgxj5yue/ on an emulator using ios 14 it opens the keyboard even on setTimeout, on iOS 16 it does not :/

**Comment by thethirdtomas** [comment]
Hello! Did this feature make it to iOS 17?

**Comment by whsieh** [comment]
> Hello! Did this feature make it to iOS 17?

It did, yes.

I think something along these lines should allow you to programmatically bring up the keyboard:

```
[_webView evaluateJavaScript:@"document.querySelector('input').focus()" completionHandler:nil];
[_webView becomeFirstResponder];
```

**Comment by mudigone** [comment]
It's not working for me. It focuses the input but the keyboard is not showing up. `inputRef.current.focus()`

**Comment by whsieh** [comment]
> It's not working for me. It focuses the input but the keyboard is not showing up. `inputRef.current.focus()`

Hm... I assume this is on iOS 17. Can you provide a simple test app (e.g. Xcode project) that demonstrates the issue?

**Comment by Sovai** [comment]
same issue for me. input is focus but keyboard not showing up. 

> It's not working for me. It focuses the input but the keyboard is not showing up. `inputRef.current.focus()`

**Comment by matvejs16** [comment]
> same issue for me. input is focus but keyboard not showing up.
> 
> > It's not working for me. It focuses the input but the keyboard is not showing up. `inputRef.current.focus()`

Can also confirm, same for me.

**Comment by whsieh** [comment]
If you can attach an Xcode project (or snippet of code) that reproduces the issue, that would be great.

To reiterate — this patch does not change the behavior of `focus()` within Safari on iOS, when called from JavaScript running on the page. It only changes the behavior of `focus()` when called from script injected via `-[WKWebView evaluateJavaScript:completionHandler:]` or one of the related methods like `-[WKWebView callAsyncJavaScript:arguments:inFrame:inContentWorld:completionHandler:]`.

In other words, web authors should not see any change in behavior — only app authors embedding `WKWebView`.

**Comment by matvejs16** [comment]
> If you can attach an Xcode project (or snippet of code) that reproduces the issue, that would be great.
> 
> To reiterate — this patch does not change the behavior of `focus()` within Safari on iOS, when called from JavaScript running on the page. It only changes the behavior of `focus()` when called from script injected via `-[WKWebView evaluateJavaScript:completionHandler:]` or one of the related methods like `-[WKWebView callAsyncJavaScript:arguments:inFrame:inContentWorld:completionHandler:]`.
> 
> In other words, web authors should not see any change in behavior — only app authors embedding `WKWebView`.

I will explain, I'm creating Telegram Mini App and I faced this issue, I think they definitely use WebView to show webpages inside their app, sorry, I'm not familiar with iOS development, so maybe this will help somehow.

**Comment by matvejs16** [comment]
For context, right now using ``iOS 18.0 (22A5297f)``

**Comment by whsieh** [comment]
> > If you can attach an Xcode project (or snippet of code) that reproduces the issue, that would be great.
> > To reiterate — this patch does not change the behavior of `focus()` within Safari on iOS, when called from JavaScript running on the page. It only changes the behavior of `focus()` when called from script injected via `-[WKWebView evaluateJavaScript:completionHandler:]` or one of the related methods like `-[WKWebView callAsyncJavaScript:arguments:inFrame:inContentWorld:completionHandler:]`.
> > In other words, web authors should not see any change in behavior — only app authors embedding `WKWebView`.
> 
> I will explain, I'm creating Telegram Mini App and I faced this issue, I think they definitely use WebView to show webpages inside their app, sorry, I'm not familiar with iOS development, so maybe this will help somehow.

I see. If the call to `focus()` comes from JavaScript from the web content (as opposed to JavaScript that's explicitly injected by the embedding application — in this case, Telegram) then I would expect the behavior to be unchanged by this patch (i.e. only programmatic focus during user-triggered events have the capability to show the keyboard).

**Comment by matvejs16** [comment]
> > > If you can attach an Xcode project (or snippet of code) that reproduces the issue, that would be great.
> > > To reiterate — this patch does not change the behavior of `focus()` within Safari on iOS, when called from JavaScript running on the page. It only changes the behavior of `focus()` when called from script injected via `-[WKWebView evaluateJavaScript:completionHandler:]` or one of the related methods like `-[WKWebView callAsyncJavaScript:arguments:inFrame:inContentWorld:completionHandler:]`.
> > > In other words, web authors should not see any change in behavior — only app authors embedding `WKWebView`.
> > 
> > 
> > I will explain, I'm creating Telegram Mini App and I faced this issue, I think they definitely use WebView to show webpages inside their app, sorry, I'm not familiar with iOS development, so maybe this will help somehow.
> 
> I see. If the call to `focus()` comes from JavaScript from the web content (as opposed to JavaScript that's explicitly injected by the embedding application — in this case, Telegram) then I would expect the behavior to be unchanged by this patch (i.e. only programmatic focus during user-triggered events have the capability to show the keyboard).

Well, then thanks for explaning and fast responses 🫡
Btw, maybe I can fill some issue or it's expected behavior?

**Comment by whsieh** [comment]
> Well, then thanks for explaning and fast responses 🫡 Btw, maybe I can fill some issue or it's expected behavior?

It's expected, insofar as it's very long-standing behavior in WebKit that's intended to reduce the possibility of content that tries to maliciously steal keyboard focus. However, it seems possible in this case for Telegram to expose some kind of API (for instance, via a `postMessage` that mini apps can call to tell Telegram to focus an element by injecting script in the web view from the application).

**Comment by spacecat** [comment]
> If you can attach an Xcode project (or snippet of code) that reproduces the issue, that would be great.
> 
> To reiterate — this patch does not change the behavior of `focus()` within Safari on iOS, when called from JavaScript running on the page. It only changes the behavior of `focus()` when called from script injected via `-[WKWebView evaluateJavaScript:completionHandler:]` or one of the related methods like `-[WKWebView callAsyncJavaScript:arguments:inFrame:inContentWorld:completionHandler:]`.
> 
> In other words, web authors should not see any change in behavior — only app authors embedding `WKWebView`.

1. When will this be available for web authors?
2. And what is the best and most simple work-around?
